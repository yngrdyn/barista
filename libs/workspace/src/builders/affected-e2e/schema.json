{
  "$schema": "http://json-schema.org/schema",
  "id": "affected-e2e",
  "type": "object",
  "properties": {
    "e2eTarget": {
      "type": "string",
      "description": "The target configuration that should be used."
    },
    "testAppSourceFolder": {
      "type": "string",
      "description": "The app root of the end to end test that should be scanned for modules using the affected components"
    }
  },
  "additionalProperties": false,
  "required": ["e2eTarget", "testAppSourceFolder"],
  "definitions": {}
}
