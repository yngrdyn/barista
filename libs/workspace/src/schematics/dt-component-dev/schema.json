{
  "$schema": "http://json-schema.org/schema",
  "id": "DtComponentDev",
  "title": "DtComponent Dev Schema",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "Name of the component",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use for your Dev component?"
    }
  }
}
