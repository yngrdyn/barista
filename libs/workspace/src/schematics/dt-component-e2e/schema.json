{
  "$schema": "http://json-schema.org/schema",
  "id": "DtComponentE2E",
  "title": "DtComponent E2E Schema",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "Name of the component",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use for your e2e component?"
    }
  }
}
