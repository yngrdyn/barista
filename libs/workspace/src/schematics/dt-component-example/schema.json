{
  "$schema": "http://json-schema.org/schema",
  "id": "DtComponentExample",
  "title": "DtComponent Example Schema",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "Name of the example",
      "default": "default",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What Name would you like to use for the example?"
    },
    "component": {
      "type": "string",
      "description": "Name of the component",
      "$default": {
        "$source": "argv",
        "index": 1
      },
      "x-prompt": "What component does the example belong to?"
    }
  }
}
