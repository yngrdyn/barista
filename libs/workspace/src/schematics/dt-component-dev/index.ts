/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { strings } from '@angular-devkit/core';
import {
  apply,
  chain,
  mergeWith,
  Rule,
  template,
  Tree,
  url,
} from '@angular-devkit/schematics';
import { join } from 'path';
import * as ts from 'typescript';
import { findNodes, getSourceFile, getIndentation } from '../utils/ast-utils';
import { commitChanges, InsertChange } from '../utils/change';
import { generateComponentOptions } from '../utils/schematics-utils';
import { DtComponentDevOptions } from './schema';
import { extend } from 'lodash-es';

interface DtDevExtendedOptions {
  selector: string;
  componentModule: {
    name: string;
    package: string;
  };
  devComponent: {
    component: string;
  };
  name: string;
}

function generateDevComponentOptions(name: string): { component: string } {
  return {
    component: `${strings.classify(name)}Demo`,
  };
}

function addImport(
  sourceFile: ts.SourceFile,
  name: string,
  modulePath: string,
): InsertChange {
  const lastImport = findNodes(
    sourceFile,
    ts.SyntaxKind.ImportDeclaration,
  ).pop() as ts.ImportDeclaration;
  const end = lastImport.end + 1;
  const dashName = strings.dasherize(name);
  const classyName = strings.classify(name);
  const toInsertImport = `import { ${classyName}Demo } from './${dashName}/${dashName}-demo.component';
  `;
  return new InsertChange(modulePath, end, toInsertImport);
}

/**
 * Adds a new route inside the demos routes
 */
export function addRoute(options: DtDevExtendedOptions): Rule {
  return (host: Tree) => {
    const modulePath = join('apps', 'dev', 'src', 'devapp-routing.module.ts');
    const sourceFile = getSourceFile(host, modulePath);

    const importChange = addImport(sourceFile, options.name, modulePath);

    const routesDeclaration = findNodes(
      sourceFile,
      ts.SyntaxKind.VariableDeclaration,
    ).find(
      (node: ts.VariableDeclaration) => node.name.getText() === 'routes',
    ) as ts.VariableDeclaration;

    const routesElements = (routesDeclaration.initializer as ts.ArrayLiteralExpression)
      .elements;
    const lastElement = routesElements[routesElements.length - 1];
    const endRoutes = routesElements.hasTrailingComma
      ? lastElement.getEnd() + 1
      : lastElement.getEnd();

    const toInsertRoute = `${getIndentation(
      routesElements,
    )}{ path: '${strings.dasherize(
      options.name,
    )}', component: ${strings.classify(options.name)}Demo },`;
    const routeChange = new InsertChange(modulePath, endRoutes, toInsertRoute);

    return commitChanges(host, [importChange, routeChange], modulePath);
  };
}

/**
 * Adds a new navigation item into demos.
 */
export function addNavItem(options: DtDevExtendedOptions): Rule {
  return (host: Tree) => {
    const modulePath = join('apps', 'dev', 'src', 'devapp.component.ts');
    const sourceFile = getSourceFile(host, modulePath);

    const navItemDeclaration = findNodes(
      sourceFile,
      ts.SyntaxKind.PropertyDeclaration,
    ).find(
      (node: ts.PropertyDeclaration) => node.name.getText() === 'navItems',
    ) as ts.PropertyDeclaration;

    const navItems = (navItemDeclaration.initializer as ts.ArrayLiteralExpression)
      .elements;
    const lastNavItem = navItems[
      navItems.length - 1
    ] as ts.ObjectLiteralExpression;
    const elementsEnd = navItems.hasTrailingComma
      ? lastNavItem.getEnd() + 1
      : lastNavItem.getEnd();
    const navChange = new InsertChange(
      modulePath,
      elementsEnd,
      `${getIndentation(navItems)}{ name: '${strings.classify(
        options.name,
      )}', route: '/${strings.dasherize(options.name)}' },`,
    );

    return commitChanges(host, [navChange], modulePath);
  };
}

/**
 * Add module.
 */
export function addModule(options: DtDevExtendedOptions): Rule {
  return (host: Tree) => {
    const modulePath = join('apps', 'dev', 'src', 'app.module.ts');
    const sourceFile = getSourceFile(host, modulePath);

    const importChange = addImport(sourceFile, options.name, modulePath);

    const moduleDeclarations = findNodes(
      sourceFile,
      ts.SyntaxKind.PropertyAssignment,
    ).find(
      (node: ts.PropertyAssignment) => node.name.getText() === 'declarations',
    ) as ts.PropertyAssignment;

    const moduleDeclarationArray = moduleDeclarations.initializer as ts.ArrayLiteralExpression;
    const moduleElements = moduleDeclarationArray.elements;
    const lastModule = moduleElements[moduleElements.length - 1];
    const modulesEnd = moduleElements.hasTrailingComma
      ? lastModule.getEnd() + 1
      : lastModule.getEnd();
    const moduleToInsert = `${getIndentation(moduleElements)}${strings.classify(
      options.name,
    )}Demo`;

    const moduleChange = new InsertChange(
      modulePath,
      modulesEnd,
      moduleToInsert,
    );

    return commitChanges(host, [importChange, moduleChange], modulePath);
  };
}

export default function (options: DtComponentDevOptions): Rule {
  return async (_host: Tree) => {
    const extendedOptions: DtDevExtendedOptions = {
      ...options,
      selector: `dt-${strings.dasherize(options.name)}`,
      componentModule: generateComponentOptions(options.name),
      devComponent: generateDevComponentOptions(options.name),
    };

    const templateSource = apply(url('./files'), [
      template({
        ...strings,
        ...extendedOptions,
      }),
    ]);

    return chain([
      mergeWith(templateSource),
      addRoute(extendedOptions),
      addNavItem(extendedOptions),
      addModule(extendedOptions),
    ]);
  };
}
