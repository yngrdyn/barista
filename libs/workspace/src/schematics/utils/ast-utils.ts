/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { SchematicsException, Tree } from '@angular-devkit/schematics';
import * as ts from 'typescript';

export function getSourceFile(host: Tree, path: string): ts.SourceFile {
  const buffer = host.read(path);
  if (!buffer) {
    throw new SchematicsException(`File ${path} does not exist.`);
  }
  const content = buffer.toString();
  return ts.createSourceFile(path, content, ts.ScriptTarget.Latest, true);
}

/**
 * Find all nodes from the AST in the subtree of node of SyntaxKind kind.
 */
export function findNodes(
  node: ts.Node,
  kind: ts.SyntaxKind,
  max: number = Infinity,
): ts.Node[] {
  if (!node || max === 0) {
    return [];
  }

  const arr: ts.Node[] = [];
  if (node.kind === kind) {
    arr.push(node);
    max--;
  }
  if (max > 0) {
    for (const child of node.getChildren()) {
      // tslint:disable-next-line: no-shadowed-variable
      findNodes(child, kind, max).forEach((node: ts.Node) => {
        if (max > 0) {
          arr.push(node);
        }
        max--;
      });

      if (max <= 0) {
        break;
      }
    }
  }

  return arr;
}

/**
 * Gets the indentation string for the last entry in NodeArray
 */
export function getIndentation(
  elements: ts.NodeArray<any> | ts.Node[],
): string {
  let indentation = '\n';
  if (elements.length > 0) {
    const text = elements[elements.length - 1].getFullText();
    const matches = text.match(/^\r?\n\s*/);
    if (matches && matches.length > 0) {
      indentation = matches[0];
    }
  }
  return indentation;
}
