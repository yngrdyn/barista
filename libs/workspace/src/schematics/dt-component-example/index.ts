/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { strings } from '@angular-devkit/core';
import {
  apply,
  chain,
  mergeWith,
  Rule,
  template,
  move,
  Tree,
  url,
  noop,
} from '@angular-devkit/schematics';
import { formatFiles } from '@nrwl/workspace';
import { DtComponentExampleOptions } from './schema';
import { normalize } from '@angular-devkit/core';
import { LICENSE_HEADER } from '../../utils/common-utils';
import { join } from 'path';
import {
  getSourceFile,
  findNodes,
  addImport,
  addExport,
} from '../utils/ast-utils';
import * as ts from 'typescript';
import { commitChanges, InsertChange } from '../utils/change';
import {
  updateExamplesModule,
  changeRoutingModule,
  changeNavigation,
  updateExamplesBarrel,
} from './example-tools.utils';

export interface DtExampleExtendedOptions {
  componentSelector: string;
  examplesModule: string;
  exampleRoute: string;
  exampleId: string;
  selector: string;
  componentModule: {
    name: string;
    package: string;
  };
  exampleComponent: {
    component: string;
    module: string;
    modulesConstant: string;
    template: string;
  };
  name: string;
  component: string;
  dashName: string;
}

const EXAMPLES_SOURCE_PATH = 'libs/examples/src/';

/**
 * Function to generate util options for the component level generation.
 * @param options
 */
function generateComponentOptions(
  options: DtComponentExampleOptions,
): { name: string; package: string } {
  return {
    name: `Dt${strings.classify(options.component)}Module`,
    package: `@dynatrace/barista-components/${strings.dasherize(
      options.component,
    )}`,
  };
}

/**
 * Function to generate util options to be used during example generation.
 * @param options
 */
function generateExampleComponentOptions(
  options: DtComponentExampleOptions,
): {
  component: string;
  module: string;
  modulesConstant: string;
  template: string;
} {
  const dashName = strings.dasherize(options.component);
  return {
    component: `DtExample${strings.classify(options.name)}${strings.classify(
      options.component,
    )}`,
    module: `DtExample${strings.classify(options.name)}Module`,
    modulesConstant: `DT_${strings
      .underscore(options.component)
      .toUpperCase()}_EXAMPLES`,
    template: `${dashName}-${strings.dasherize(options.name)}-example.html`,
  };
}

/**
 * Generate the index file for the examples.
 * @param options
 */
function generateBarrelFileContent(options: DtExampleExtendedOptions): string {
  return `${LICENSE_HEADER}

  export * from './${options.dashName}-examples.module';
  export * from './${options.exampleRoute}';
  `;
}

/**
 * Generate the module file for the examples.
 * @param options
 */
function generateModuleFileContent(options: DtExampleExtendedOptions): string {
  return `${LICENSE_HEADER}
  import { NgModule } from '@angular/core';
  import { ${options.componentModule.name} } from '${options.componentModule.package}';
  import { ${options.exampleComponent.component} } from './${options.exampleRoute}';

  @NgModule({
    imports: [${options.componentModule.name}],
    declarations: [${options.exampleComponent.component}],
  })
  export class ${options.examplesModule} {}`;
}

/**
 * Function to update the module files.
 * @param options
 */
function updateModules(options: DtExampleExtendedOptions): Rule {
  return (host: Tree) => {
    try {
      // add import
      const modulePath = join(
        'libs',
        'examples',
        'src',
        options.dashName,
        `${options.dashName}-examples.module.ts`,
      );
      const sourceFile = getSourceFile(host, modulePath);

      const importChange = addImport(
        sourceFile,
        options.exampleComponent.component,
        options.exampleRoute,
        modulePath,
      );
      // find last module in the array, and add new module
      const modulesDeclaration = findNodes(
        sourceFile,
        ts.SyntaxKind.PropertyAssignment,
      ).find(
        (node: ts.PropertyAssignment) => node.name.getText() === 'declarations',
      ) as ts.PropertyAssignment;
      const modulesElements = (modulesDeclaration.initializer as ts.ArrayLiteralExpression)
        .elements;
      const lastElement = modulesElements[modulesElements.length - 1];
      const end = modulesElements.hasTrailingComma
        ? lastElement.getEnd() + 1
        : lastElement.getEnd();
      const routesChange = new InsertChange(
        modulePath,
        end,
        `${modulesElements.hasTrailingComma ? ' ' : ', '}${
          options.exampleComponent.component
        }`,
      );

      return commitChanges(host, [importChange, routesChange], modulePath);
    } catch (error) {
      throw new Error(
        `There was a problem updating ${options.dashName}-examples.module.ts ${error.message}`,
      );
    }
  };
}

/**
 * Function to update the index file.
 * @param options
 */
function updateIndex(options: DtExampleExtendedOptions): Rule {
  return (host: Tree) => {
    try {
      // add export to index.ts
      const indexPath = join(
        'libs',
        'examples',
        'src',
        options.dashName,
        'index.ts',
      );
      const sourceIndexFile = getSourceFile(host, indexPath);
      const exportChange = addExport(
        sourceIndexFile,
        options.exampleRoute,
        indexPath,
      );

      return commitChanges(host, exportChange, indexPath);
    } catch (error) {
      throw new Error(
        `There was a problem updating 'libs/examples/src/${options.dashName}/index.ts' : ${error.message}`,
      );
    }
  };
}

/**
 * The main function of the example schematics.
 * @param options
 */
export default function (options: DtComponentExampleOptions): Rule {
  return async (tree: Tree) => {
    const dashName = strings.dasherize(options.component);
    const exampleId = `${dashName}-${strings.dasherize(options.name)}`;
    const extendedOptions: DtExampleExtendedOptions = {
      ...options,
      componentSelector: `dt-${dashName}`,
      selector: `dt-example-${strings.dasherize(options.name)}-${dashName}`,
      componentModule: generateComponentOptions(options),
      exampleComponent: generateExampleComponentOptions(options),
      examplesModule: `DtExamples${strings.classify(options.component)}Module`,
      exampleRoute: `${exampleId}-example/${exampleId}-example`,
      exampleId,
      dashName: dashName,
    };

    // Generating files from file templates, and moving them to their path.
    const templateSource = apply(url('./files'), [
      template({
        ...strings,
        ...extendedOptions,
      }),
      move(
        normalize(
          `${EXAMPLES_SOURCE_PATH}${dashName}/${dashName}-${strings.dasherize(
            options.name,
          )}-example`,
        ),
      ),
    ]);

    const indexPath = normalize(`${EXAMPLES_SOURCE_PATH}${dashName}/index.ts`);

    const isNewComponent = !tree.exists(indexPath);

    // If the example is for a new component, so there are no other examples for
    // for this component yet, we have to generate the index and module files.
    if (isNewComponent) {
      tree.create(indexPath, generateBarrelFileContent(extendedOptions));

      const modulePath = normalize(
        `${EXAMPLES_SOURCE_PATH}${dashName}/${dashName}-examples.module.ts`,
      );
      tree.create(modulePath, generateModuleFileContent(extendedOptions));
    }

    return chain([
      mergeWith(templateSource),
      isNewComponent ? noop() : updateModules(extendedOptions),
      isNewComponent ? noop() : updateIndex(extendedOptions),
      isNewComponent ? updateExamplesModule(extendedOptions) : noop(),
      updateExamplesBarrel(extendedOptions, isNewComponent),
      changeRoutingModule(extendedOptions),
      changeNavigation(extendedOptions, isNewComponent),
      formatFiles(),
    ]);
  };
}
