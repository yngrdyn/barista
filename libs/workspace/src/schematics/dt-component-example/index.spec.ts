/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Tree } from '@angular-devkit/schematics';
import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import {
  Schema as ApplicationOptions,
  Style,
} from '@schematics/angular/application/schema';
import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
import { join } from 'path';
import { addFixtureToTree, getNewFiles } from '../utils/test-utils';

const fixturesFolder = join(__dirname, './fixtures');

describe('Add Example Component', () => {
  const workspaceOptions: WorkspaceOptions = {
    name: 'workspace',
    newProjectRoot: 'apps',
    version: '8.0.0',
  };

  const appOptions: ApplicationOptions = {
    name: 'myapp',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: Style.Scss,
    skipTests: true,
    skipPackageJson: true,
  };

  let initialTree: UnitTestTree;
  let schematicRunner: SchematicTestRunner;
  let initialFiles: string[];

  beforeEach(async () => {
    schematicRunner = new SchematicTestRunner(
      '@dynatrace/barista-components/schematics',
      join(__dirname, '../../../collection.json'),
    );

    initialTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions,
        Tree.empty(),
      )
      .toPromise();

    initialTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        initialTree,
      )
      .toPromise();

    await addFixtureToTree(
      initialTree,
      'examples.module.ts.fixture',
      '/libs/examples/src/examples.module.ts',
      fixturesFolder,
    );

    await addFixtureToTree(
      initialTree,
      'examplesIndex.ts.fixtures',
      '/libs/examples/src/index.ts',
      fixturesFolder,
    );

    await addFixtureToTree(
      initialTree,
      'app-routing.module.ts.fixture',
      '/apps/demos/src/app-routing.module.ts',
      fixturesFolder,
    );

    await addFixtureToTree(
      initialTree,
      'nav-items.ts.fixture',
      '/apps/demos/src/nav-items.ts',
      fixturesFolder,
    );

    initialFiles = initialTree.files;
  });

  it('should add the necessary files', async () => {
    const result = await schematicRunner
      .runSchematicAsync(
        'dt-example',
        { name: 'simple', component: 'TestComponent' },
        initialTree,
      )
      .toPromise();

    const newFiles = getNewFiles(result.files, initialFiles);

    expect(newFiles).toEqual([
      '/libs/examples/src/test-component/index.ts',
      '/libs/examples/src/test-component/test-component-examples.module.ts',
      '/libs/examples/src/test-component/test-component-simple-example/test-component-simple-example.html',
      '/libs/examples/src/test-component/test-component-simple-example/test-component-simple-example.ts',
    ]);
  });

  it('should generate files with the appropriate content', async () => {
    const result = await schematicRunner
      .runSchematicAsync(
        'dt-example',
        { name: 'simple', component: 'TestComponent' },
        initialTree,
      )
      .toPromise();

    const newFiles = getNewFiles(result.files, initialFiles);
    newFiles.map((file) => expect(result.readContent(file)).toMatchSnapshot());
  });

  it('should change files appropriately', async () => {
    await addFixtureToTree(
      initialTree,
      'test-component-examples.module.ts.fixture',
      '/libs/examples/src/test-component/test-component-examples.module.ts',
      fixturesFolder,
    );

    await addFixtureToTree(
      initialTree,
      'index.ts.fixture',
      '/libs/examples/src/test-component/index.ts',
      fixturesFolder,
    );

    const result = await schematicRunner
      .runSchematicAsync(
        'dt-example',
        { name: 'complex', component: 'TestComponent' },
        initialTree,
      )
      .toPromise();

    expect(
      result.readContent(
        '/libs/examples/src/test-component/test-component-examples.module.ts',
      ),
    ).toMatchSnapshot();

    expect(
      result.readContent('/libs/examples/src/test-component/index.ts'),
    ).toMatchSnapshot();

    expect(result.readContent('/libs/examples/src/index.ts')).toMatchSnapshot();

    expect(
      result.readContent('/libs/examples/src/examples.module.ts.ts'),
    ).toMatchSnapshot();

    expect(
      result.readContent('/apps/demos/src/app-routing.module.ts'),
    ).toMatchSnapshot();

    expect(
      result.readContent('/apps/demos/src/nav-items.ts'),
    ).toMatchSnapshot();
  });

  it('should throw error if we already have the files in there', async () => {
    await addFixtureToTree(
      initialTree,
      'test-test-example.fixture',
      '/libs/examples/src/test/test-test-example/test-test-example.ts',
      fixturesFolder,
    );

    await expect(
      schematicRunner
        .runSchematicAsync(
          'dt-example',
          { name: 'test', component: 'test' },
          initialTree,
        )
        .toPromise(),
    ).rejects.toThrow();
  });

  it('should throw error if there was a problem updating examples.module file', async () => {
    await addFixtureToTree(
      initialTree,
      'faulty.fixture',
      '/libs/examples/src/test/test-test-example/test-test-example.ts',
      fixturesFolder,
    );

    await expect(
      schematicRunner
        .runSchematicAsync(
          'dt-example',
          { name: 'test', component: 'test' },
          initialTree,
        )
        .toPromise(),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should provide meaningful error message if component index.ts modification is not successful', async () => {
    await addFixtureToTree(
      initialTree,
      'test-component-examples.module.ts.fixture',
      '/libs/examples/src/test-component/test-component-examples.module.ts',
      fixturesFolder,
    );

    await addFixtureToTree(
      initialTree,
      'faulty.fixture',
      '/libs/examples/src/test-component/index.ts',
      fixturesFolder,
    );

    await expect(
      schematicRunner
        .runSchematicAsync(
          'dt-example',
          { name: 'complex', component: 'TestComponent' },
          initialTree,
        )
        .toPromise(),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should provide meaningful error message if component nav-items.ts modification is not successful', async () => {
    await addFixtureToTree(
      initialTree,
      'faulty.fixture',
      '/apps/demos/src/nav-items.ts',
      fixturesFolder,
    );

    await expect(
      schematicRunner
        .runSchematicAsync(
          'dt-example',
          { name: 'complex', component: 'TestComponent' },
          initialTree,
        )
        .toPromise(),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should provide meaningful error message if component app-routing.module.ts modification is not successful', async () => {
    await addFixtureToTree(
      initialTree,
      'faulty.fixture',
      '/apps/demos/src/app-routing.module.ts',
      fixturesFolder,
    );

    await expect(
      schematicRunner
        .runSchematicAsync(
          'dt-example',
          { name: 'complex', component: 'TestComponent' },
          initialTree,
        )
        .toPromise(),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should provide meaningful error message if barrel index.ts modification is not successful', async () => {
    await addFixtureToTree(
      initialTree,
      'faulty.fixture',
      '/libs/examples/src/index.ts',
      fixturesFolder,
    );

    await expect(
      schematicRunner
        .runSchematicAsync(
          'dt-example',
          { name: 'complex', component: 'TestComponent' },
          initialTree,
        )
        .toPromise(),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
