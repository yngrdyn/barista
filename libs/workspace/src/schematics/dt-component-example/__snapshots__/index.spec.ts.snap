// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Example Component should change files appropriately 1`] = `
"/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the \\"License\\");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an \\"AS IS\\" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { NgModule } from '@angular/core';
import { DtTestComponentModule } from '@dynatrace/barista-components/test-component';
import { DtExampleSimpleTestComponent } from './test-component-simple-example/test-component-simple-example';
import { DtExampleComplexTestComponent } from './test-component-complex-example/test-component-complex-example';
export const DT_TEST_COMPONENT_EXAMPLES = [
  DtExampleSimpleTestComponent,
  DtExampleComplexTestComponent,
];

@NgModule({
  imports: [DtTestComponentModule],
  declarations: [...DT_TEST_COMPONENT_EXAMPLES],
  entryComponents: [...DT_TEST_COMPONENT_EXAMPLES],
})
export class DtExamplesTestComponentModule {}
"
`;

exports[`Add Example Component should change files appropriately 2`] = `
"/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the \\"License\\");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an \\"AS IS\\" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export * from './test-component-examples.module';
export * from './test-component-simple-example/test-component-simple-example';
export * from './test-component-complex-example/test-component-complex-example';
"
`;

exports[`Add Example Component should change files appropriately 3`] = `
"import { Type } from '@angular/core';

import { DtExampleAlertDarkError } from './alert/alert-dark-error-example/alert-dark-error-example';
import { DtExampleAlertDark } from './alert/alert-dark-example/alert-dark-example';
import { DtExampleAlertError } from './alert/alert-error-example/alert-error-example';
import { DtExampleComplexTestComponent } from '././test-component/test-component-complex-example/test-component-complex-example';
export { DtExamplesModule } from './examples.module';
export { DtExamplesTreeTableModule } from './tree-table/tree-table-examples.module';
export {
  DtExampleAlertDarkError,
  DtExampleAlertDark,
  DtExampleTreeTableSimple,
  DtExampleComplexTestComponent,
};

export const EXAMPLES_MAP = new Map<string, Type<unknown>>([
  ['DtExampleAlertDarkError', DtExampleAlertDarkError],
  ['DtExampleAlertDark', DtExampleAlertDark],
  ['DtExampleComplexTestComponent', DtExampleComplexTestComponent],
]);
"
`;

exports[`Add Example Component should change files appropriately 4`] = `""`;

exports[`Add Example Component should change files appropriately 5`] = `
"import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';

import {
  DtExampleTreeTableProblemIndicator,
  DtExampleTreeTableSimple,
  DtExampleComplexTestComponent,
} from '@dynatrace/examples';

const ROUTES: Routes = [
  { path: 'alert-dark-error-example', component: DtExampleAlertDarkError },
  { path: 'tree-table-default-example', component: DtExampleTreeTableDefault },
  {
    path: 'tree-table-problem-indicator-example',
    component: DtExampleTreeTableProblemIndicator,
  },
  { path: 'tree-table-simple-example', component: DtExampleTreeTableSimple },
  {
    path: 'test-component-complex-example',
    component: DtExampleComplexTestComponent,
  },
];

@NgModule({
  imports: [RouterModule.forRoot(ROUTES)],
  exports: [RouterModule],
})
export class DtDemosAppRoutingModule {}
"
`;

exports[`Add Example Component should change files appropriately 6`] = `
"export const DT_DEMOS_EXAMPLE_NAV_ITEMS = [
  {
    name: 'alert',
    examples: [
      {
        name: 'alert-dark-error-example',
        route: '/alert-dark-error-example',
      },
    ],
  },
  {
    name: 'test-component',
    examples: [
      {
        name: 'autocomplete-attach-different-element-example',
        route: '/autocomplete-attach-different-element-example',
      },
      {
        name: 'test-component-complex-example',
        route: '/test-component-complex-example',
      },
    ],
  },
];
"
`;

exports[`Add Example Component should generate files with the appropriate content 1`] = `
"/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the \\"License\\");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an \\"AS IS\\" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export * from './test-component-examples.module';
export * from './test-component-simple-example/test-component-simple-example';
"
`;

exports[`Add Example Component should generate files with the appropriate content 2`] = `
"/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the \\"License\\");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an \\"AS IS\\" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { NgModule } from '@angular/core';
import { DtTestComponentModule } from '@dynatrace/barista-components/test-component';
import { DtExampleSimpleTestComponent } from './test-component-simple-example/test-component-simple-example';

export const DT_TEST_COMPONENT_EXAMPLES = [DtExampleSimpleTestComponent];

@NgModule({
  imports: [DtTestComponentModule],
  declarations: [...DT_TEST_COMPONENT_EXAMPLES],
  entryComponents: [...DT_TEST_COMPONENT_EXAMPLES],
})
export class DtExamplesTestComponentModule {}
"
`;

exports[`Add Example Component should generate files with the appropriate content 3`] = `
"<dt-test-component></dt-test-component>
"
`;

exports[`Add Example Component should generate files with the appropriate content 4`] = `
"/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the \\"License\\");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an \\"AS IS\\" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component } from '@angular/core';

@Component({
  selector: 'dt-example-simple-test-component',
  templateUrl: './test-component-simple-example.html',
})
export class DtExampleSimpleTestComponent {}
"
`;
